packages/core/src/pubsub/incoming-channel.ts

```diff
@@ -0,0 +1,107 @@
+import { interval, Observable } from 'rxjs';
+import { Memoize } from 'typescript-memoize';
+import { IpfsApi } from '@ceramicnetwork/common';
+import { TaskQueue } from './task-queue';
+import { DiagnosticsLogger, ServiceLogger } from '@ceramicnetwork/logger';
+import { pipe, MonoTypeOperatorFunction } from 'rxjs';
+import { map, filter, concatMap } from 'rxjs/operators';
+
+// Typestub for pubsub message.
+// At some future time this type definition should be provided by IPFS.
+export type IPFSPubsubMessage = {
+  from: string;
+  seqno: Uint8Array;
+  data: Uint8Array;
+  topicIDs: string[];
+  signature: Uint8Array;
+  key: Uint8Array;
+};
+
+async function resubscribe(
+  ipfs: IpfsApi,
+  topic: string,
+  pubsubLogger: ServiceLogger,
+  handler: (message: IPFSPubsubMessage) => void,
+) {
+  const listeningTopics = await ipfs.pubsub.ls();
+  const isSubscribed = listeningTopics.includes(topic);
+  if (!isSubscribed) {
+    await ipfs.pubsub.unsubscribe(topic, handler);
+    await ipfs.pubsub.subscribe(topic, handler);
+    const ipfsId = await ipfs.id();
+    const peerId = ipfsId.id;
+    pubsubLogger.log({ peer: peerId, event: 'subscribed', topic: topic });
+  }
+}
+
+/**
+ * Incoming IPFS PubSub message stream as Observable.
+ * Ensures that ipfs subscription to a topic is live by periodically re-subscribing.
+ *
+ * One could simplify this by removing periodic re-subscription and serialization via task-queue.
+ * Unfortunately, we can end up with concurrency issues: trying to unsubscribe while (or right after) other instance tries to subscribe.
+ * So, better keep it all together.
+ */
+export class IncomingChannel extends Observable<IPFSPubsubMessage> {
+  constructor(
+    ipfs: IpfsApi,
+    topic: string,
+    resubscribeEvery: number,
+    pubsubLogger: ServiceLogger,
+    readonly logger: DiagnosticsLogger,
+  ) {
+    super((subscriber) => {
+      const handler = (message: IPFSPubsubMessage) => subscriber.next(message);
+
+      this.tasks.add(() => resubscribe(ipfs, topic, pubsubLogger, handler));
+
+      const ensureSubscribed = interval(resubscribeEvery).subscribe(() => {
+        this.tasks.add(() => resubscribe(ipfs, topic, pubsubLogger, handler));
+      });
+
+      return () => {
+        ensureSubscribed.unsubscribe();
+        // Remove pending subscription attempts.
+        this.tasks.clear();
+        // Unsubscribe only after a currently running task is finished.
+        this.tasks.add(async () => {
+          await ipfs.pubsub.unsubscribe(topic, handler);
+        });
+      };
+    });
+  }
+
+  // We want all the tasks added to execute in FIFO order.
+  // Subscription attempts must be sequential.
+  // Last call to unsubscribe must execute after all the attempts are done,
+  // and all the attempts yet inactive are cleared.
+  @Memoize()
```

Comment:
This is a little more magical/indirect than I prefer.  Can we just make a private `#tasks` member that we initialize once in the constructor, rather than memoizing the getter like we're doing here?

Reply:
Agree on magical/indirect. All the other ways I can think of to structure this are even more indirect an definitely less clear: https://gist.github.com/ukstv/acebcddbf77a330f8f2d0d340a2a7910 

Maybe https://gist.github.com/ukstv/acebcddbf77a330f8f2d0d340a2a7910#file-4-ts is okay to you?

Reply:
I kind of like a mix of versions 3 and 4.  Constructing `tasks` within the body of the constructor like in version 3, but using a helper function like exists in version 4 (although I'd probably name it `makeTaskQueue` or `makeResubscribeQueue`).

Not a huge deal either way though, so go with whatever you think is cleanest

Reply:
I also like how version 3 pulls out `onSubscribe` into a helper function too, to make the constructor body much smaller and easier to digest